src/zcash/IncrementalMerkleTree.cpp

```diff
@@ -25,28 +20,41 @@ SHA256Compress SHA256Compress::combine(const SHA256Compress& a, const SHA256Comp
     return res;
 }
 
-
-template <typename Hash>
+template <size_t Depth, typename Hash>
 class PathFiller {
 private:
     std::deque<Hash> queue;
 public:
     PathFiller() : queue() { }
     PathFiller(std::deque<Hash> queue) : queue(queue) { }
 
-    Hash next() {
+    Hash next(size_t depth) {
         if (queue.size() > 0) {
             Hash h = queue.front();
             queue.pop_front();
 
             return h;
         } else {
-            return Hash();
+            return IncrementalMerkleTree<Depth, Hash>::rootOfEmptyTree(depth);
         }
     }
+
 };
 
 template<size_t Depth, typename Hash>
+Hash IncrementalMerkleTree<Depth, Hash>::rootOfEmptyTree(size_t depth) {
+    static boost::optional<boost::array<Hash, Depth+1>> empty_roots;
+    if (!empty_roots) {
+        empty_roots = boost::array<Hash, Depth+1>();
+        empty_roots->at(0) = Hash();
+        for (size_t d = 1; d <= depth; d++) {
```

Comment:
BUG: This should be `Depth` not `depth`! For example, call it with `depth=0` once and foreverafter it will return zero for any valid depth you ask for. Good thing I'm writing tests for this.